measure

```diff
@@ -87,6 +100,113 @@ insights_session = url_session(INSIGHTS_API_URL.format(account_id=NEWRELIC_ACCOU
 insights_session.headers.update({'X-Query-Key': NEWRELIC_INSIGHTS_QUERY_KEY})
 
 
+def aggregate_metric_values_to_scalar(
+    metric_name: str,
+    metrics: Dict[str, Any],
+    result: Dict[str, Any],
+    is_threshold: bool,
+) -> bool:
+    """Recieves information about a single slo condition as well as the fetched metrics and attempts to produce a scalar
+    value computed by producing a list of means from timeseries values returned per instance and producing a scalar from the mean
+    of per instance means
+
+    Args:
+        metric_name (str): Name of the metric to produce a scalar for
+        metrics (Dict[str, Any]): Fetched metrics as returned by get_metrics
+        result (Dict[str, Any]): A reference to a dictionary tracking the result of an individual condition, 
+            used to return the scalar value (keyed metric_scalar/threshold_metric_scalar), information about the metrics payload, 
+            and a message if unable to compute a valid scalar
+        is_threshold (bool): Used to format result info based on whether marshalling a condition metric or threshold_metric
+
+    Returns:
+        bool: True if a scalar was produced, false to indicate the condition relying on the metric must be skipped
+    """
+    preamble = 'Condition threshold' if is_threshold else 'Condition'
+    key_preamble = "threshold_" if is_threshold else ""
+
+    if metric_name not in metrics:
+        result['message'] = f'{preamble} metric not contained in fetched metrics'
+        return False
+
+    metric_values = metrics[metric_name]['values']
+    if not metric_values:
+        result['message'] = f'{preamble} metric contained no instances or values'
+        return False
+
+    result[f'{key_preamble}instances'] = ', '.join([val['id'] for val in metric_values])
+    instance_values = [val['data'] for val in metric_values]
+    
+    # Check at least one list has values
+    if not any(instance_values):
+        result['message'] = f'{preamble} metric instance values all had empty data'
+        return False
+
+    inst_scalars = []
+    for val_list in instance_values:
+        if not val_list:
+            continue
+
+        # Pull metric values from timestamps, filter out falsey values; 0, 0.0
+        single_inst_vals = [decimal.Decimal(v[1]) for v in val_list if v[1]] # TODO: NaN check?
```

Comment:
Why are zero values simply skipped over? AFAIK, metrics may legitimately have zero value (e.g., error count). 

Not even a negative value should be assumed invalid or impossible: there may well be measurable things that have a valid range like -1000 ... 1000, or similar.

(rant: "Falsey" is a word? I must be getting old. As is my browser's spell checker - it highlighted this as a typo. Anyway, metrics are by definition numeric, might as well use 'zero value' in the message text)

Reply:
Per the very last bullet point of the spec doc under Unresolved Issues, discerning 0 values from missing values is still being worked out. I did some digging and found if there is no data for a metric in a given timeslice, NewRelic returns 0.  Per the [docs](https://docs.newrelic.com/docs/apis/rest-api-v2/basic-functions/extract-metric-timeslice-data/#non-existent_data_help)

> Metric values that have existed in the past, but are no longer collected, will return a zero value.

I was also able to find some discussion on this in the community hub. This [particular issue](https://discuss.newrelic.com/t/api-returns-values-of-0-instead-of-null-when-using-average-function-on-a-time-period-with-no-data-points/6467/3?u=ben108) presents a solution in checking for a metric that should never be zero and using that to differentiate between missing data and true zero. In our case, we could define a `sentinel_metric` configuration within applicable Conditions for which true zero is a value that we must include in the calculations

https://discuss.newrelic.com/t/feature-idea-timeseries-queries-with-sparse-data-points/34591/13
https://discuss.newrelic.com/t/apm-layers-in-graphs-take-sharp-plunges-whilst-waiting-on-fresh-data/31835
https://discuss.newrelic.com/t/time-series-no-data-instead-of-0/45807